### ISSUE TITLE
network do not learn anything.

### ISSUE THREAD

**Issue by wrluo** [Microsoft/CNTK#584]
I have my own defined loss function which is

**loss = Times(labels, activation)**

Can I put the above loss function in NDL file directly and set

**CriterionNodes(loss)**?

In my project, it seems that learnable parameter do not change at all no matter how many epochs I take.

**Comment by frankseide** [comment]
Yes. To be precise, the following should work:

    loss = TransposeTimes (labels, activations)
    criterionNodes = (loss)

or you just say

    loss = TransposeTimes (labels, activations, tag="criterion")

In case your input is a sequence, this will give you a sequence of per-sample criterion values. Those will automatically be summed up by CNTK across all sequences and samples in each minibatch.

For example, the same loss-function pattern is used in the sequence-to-sequence example:

    ce   = Pass (ReduceLogSum (z) - TransposeTimes (labelSequence,          z),  tag='criterion')
    errs = Pass (BS.Constants.One - TransposeTimes (labelSequence, Hardmax (z)), tag='evaluation')

(Where `Pass()` just wraps a dummy node around the value. You won't need that if you say `tag="criterion"` in the `TransposeTimes()` operation.)

**Comment by wrluo** [comment]
@frankseide thanks for the help!

Actually `activations` and `labels` are column vectors. I want to minimize `(activations * labels) - 0`. I corrected my NDL file according to your suggestion, but still had no luck to get my network trained. Seems that error is regarded as 0 during backpropagation. 

Here is part of my NDL file:

```
features = Input(featDim)
labels = Input(labelDim)

LSTMoutput1 = LSTMPComponent(basefeatDim, L1_hiddenDim, L1_cellDim, features, L1_cellDimX2, L1_cellDimX3, L1_cellDimX4);

W = Parameter(labelDim, L1_hiddenDim, init="uniform", initValueScale=1);
b = Parameter(labelDim, 1,         init="fixedValue", value=0);
LSTMoutputW = Plus(Times(W, LSTMoutput1), b);

activation = Softmax(LSTMoutputW);
loss = TransposeTimes(labels, activation);

FeatureNodes = (features)
LabelNodes = (labels)
CriterionNodes = (loss)
EvalNodes = (loss, W, b, LSTMoutput1.output)
OutputNodes = (activation, W, b, LSTMoutput1.output)
```

Any idea of this?

**Comment by frankseide** [comment]
CNTK minimizes loss functions (lower=better). Your objective seem to be attempting to maximize (higher=better). Would you mind trying

    negActivation = Negate (activation)
    loss = TransposeTimes(labels, negActivation)

(BTW, no need for the semicolons, unless you want more than one expression on a single line.)

**Comment by wrluo** [comment]
@frankseide thanks for the patience~

I change the NDL file according to your suggestion, but still no luck to get the network trained. All those parameters do not change at all not matter how many epoch I take.

>     features = Input(featDim)
>     labels = Input(labelDim)
>     # labels = Softmax(Input(labelDim))
> 
>     LSTMoutput1 = LSTMComponent(basefeatDim, L1_hiddenDim, features)
> 
>     W = Parameter(labelDim, L1_hiddenDim, init="uniform", initValueScale=1)
>     b = Parameter(labelDim, 1,         init="fixedValue", value=0)
>     LSTMoutputW = Plus(Times(W, LSTMoutput1), b)
> 
>     activation = Softmax(LSTMoutputW)
>     negActivation = Negate (activation)
>     loss = TransposeTimes(labels, negActivation)
>     # loss = CrossEntropyWithSoftmax(labels, LSTMoutputW)
> 
>     FeatureNodes = (features)
>     LabelNodes = (labels)
>     CriterionNodes = (loss)
>     EvalNodes = (loss, LSTMoutput1.output)
>     OutputNodes = (activation, W, b, LSTMoutput1.output)

BTW, if I use the two commented code which are 
`# labels = Softmax(Input(labelDim))` 
and `# loss = CrossEntropyWithSoftmax(labels, LSTMoutputW)` 
to replace 
`labels = Input(labelDim)` 
and  `loss = TransposeTimes(labels, negActivation)` respectively (shown in above code),
the network's parameters will be updated perfectly during the training. This confuse me, I have no idea of what's going on of my objective. What I want is just minimize `labels*activation` .Any suggestion?

**Comment by wrluo** [comment]
CNTK works well after I set the environment variable ACML_FMA=0, I will close the issue. Thanks for your help.
